swarm/metrics.py

```diff
@@ -0,0 +1,165 @@
+
+
+def rms_error(swarm_ypreds: np.ndarray, y: np.ndarray) -> np.ndarray:
+    """
+    This is a function that calculates the root-mean sqared error for a prediction array, given true y.
+    It averages along the all the point predictions in x for a set of predictions.
+    Args:
+        swarm_ypreds: A numpy array of predictions. This can have different numbers of dimensions,
+            eg [bee, epoch, x_step] or [epoch, x_step], so long as the final dimension represents the
+            predictions at different x values, and has same length as y.
+        y: The true values predictions. A 1-d array.
+
+    Returns: A numpy array of dimension one less than the input `swarm_ypreds`, with the evaluated root-mean-squared
+    loss across all predictions.
+    """
+    dims = len(swarm_ypreds.shape)
+    squares = np.square(np.subtract(swarm_ypreds, y))
+    losses = np.apply_along_axis(np.mean, dims-1, squares)
+    return losses
+
+
+def iteration_finder(epoch_scores: [np.ndarray, list],
+                     threshold: float,
+                     threshold_type: str = "nominal",
+                     criterion: str = "first",
+                     direction: str = "below",
+                     reference_iteration: int = 0) -> [int, None]:
+    """
+    This is a function intended to find when a certain score of interest is achieved during a training process.
+    A single vector of scores representing the assessment of the whole swarm at a given epoch/batch (ie iteration)
+    is required.
+
+    In a typical case, you might want to find when a loss is below a suitable threshold, eg .05.
+    An alternative might be that you want to find when the range of standard deviations at all x values has
+    falled below a threshold. (This could indicate knots/nodes eliminated.) Or the median/mean/max standard deviation
+    has fallen below a threshold. (This would indicate a level of coherence within the swarm.)
+
+    The threshold can have two different interpretations, either as a nominal value or a ratio relative to a reference
+    value at a given iteration (often the first).  For different regimes of target functions the nominal evaluation score
+    may vary widely widely in terms of magnitude. So the threshold 0.05 can be used to indicate that the score reaches
+    5% of whatever the score was at the first epoch/batch. For ascending scores, a value greater than one may be
+    helpful. i.e. 20 would indicate improving starting value 20 times over.
+
+    This function also allows you to set a criterion, based on whether it should return the first epoch/batch at which
+    the threshold was crossed, or the first time it was crossed and never thereafter exceeded. (I.e. always remains
+    below that threshold in all tested epochs/batches.)
+    Args:
+        epoch_scores: A list or 1-d Numpy array of floats representing the scores at each epoch/catch.
```

Comment:
I'd consider doing this across `bees` - most of what you've got here can automatically be vectorised and would be faster than sticking it into a for loop

Reply:
Not sure what you mean here?  The code references don't seem to refer to a loop?  ( think I've used relatively few?)

Reply:
Oh, I meantt hat instead of `epoch_scores` being 1d, `epoch_scores` should be 2d [bee, epoch] and the function should run over that. Otherwise I'd have to do `for bee in range(epoch_scores.shape[0]): ...` which'd be slower. 
Your function might just work if it's 2d with no changes (except a return type haha)

Reply:
Hmmm, no I don't think I want `epoch_scores` to be any more than 1d. The whole point of this exercise is to make assessments of the performance of the swarm, rather than individual bees. The existing loss/score is produced for every bee. Any number of alternative per-bee metrics could be added to your (suitable open and flexible) `core.swarm_train` by adjusting the `bee_trainer`. I'm deliberately trying to track just a metric what we can learn from the entire swarm at each step.  That could be performance of the best bee or worst bee, or the performance of the aggregate prediction of all the bees together.  (The main object of the `summarise_across` function used in conjunction with `rms_error`). Maybe I need an end-to-end example of the workflow to make this clearer. But if I wanted another score on a per-bee basis, the right place to do that is to build it into the `bee-trainer` in my opinion.